src/components/Suspense.astro

```diff
@@ -1,7 +1,72 @@
 ---
-const idx = Astro.locals.suspend(Astro.slots.render("default"));
+type PendingThenable<T> = { status: "pending" };
+type FulfilledThenable<T> = { status: "fulfilled"; value: T };
+type RejectedThenable<T> = { status: "rejected"; reason: unknown };
+
+type Thenable<T> = Promise<T> &
+  (PendingThenable<T> | FulfilledThenable<T> | RejectedThenable<T>);
+
+function instrument<T>(promise: Promise<T>): Thenable<T> {
+  const thenable = promise as Promise<T> & PendingThenable<T>;
+  thenable.status = "pending";
+  thenable.then(
+    (value) => {
+      const fulfilled = promise as Promise<T> & FulfilledThenable<T>;
+      fulfilled.status = "fulfilled";
+      fulfilled.value = value;
+    },
+    (error) => {
+      const rejected = promise as Promise<T> & RejectedThenable<T>;
+      rejected.status = "rejected";
+      rejected.reason = error;
+    }
+  );
+  return thenable;
+}
+
+function sleep(ms: number) {
+  return new Promise<void>((resolve) => setTimeout(resolve, ms));
+}
+
+const promise = Astro.slots.render("default");
+const thenable = instrument(promise); // add `.status` and `.value` / `.reason`
+
+const DEADLINE_MS = 5; // arbitrary, might be too high
+await sleep(DEADLINE_MS);
+
+let idx = -1;
+let content = "";
+
+switch (thenable.status) {
+  case "pending": {
+    idx = Astro.locals.suspend(promise);
+    break;
+  }
+  case "rejected": {
+    throw thenable.reason;
```

Comment:
this'll also fail the whole request synchronously. raises the question of if/how we wanna handle errors in these -- is this fine to do? do we want a matching ErrorBoundary thing? i'm familiar with Astro's conventions around this